### PR TITLE
RavenDB-14094: Fix JsonContextPool leak (originally identified in RunLowMemoryHandlers)

### PR DESCRIPTION
**Pull request by aviviadi**

### CODE REVIEW

File: src/Raven.Server/Utils/ReplicationUtils.cs [hunk 1]
```diff
@@ -13,6 +15,35 @@ namespace Raven.Server.Utils
 {
     internal static class ReplicationUtils
     {
+        private static readonly ConcurrentDictionary<string, (X509Certificate2, ClusterRequestExecutor)> _requestExecutorsCache = new ConcurrentDictionary<string, (X509Certificate2, ClusterRequestExecutor)>();
+
+        private static ClusterRequestExecutor GetOrCreateForSingleNode(string url, X509Certificate2 certificate)
+        {
+            ClusterRequestExecutor reqExecutor;
+            if (_requestExecutorsCache.ContainsKey(url))
```
Comment (ayende) [inline src/Raven.Server/Utils/ReplicationUtils.cs]: TryGet, not Contains

File: src/Raven.Server/Utils/ReplicationUtils.cs [hunk 2]
```diff
@@ -13,6 +15,35 @@ namespace Raven.Server.Utils
 {
     internal static class ReplicationUtils
     {
+        private static readonly ConcurrentDictionary<string, (X509Certificate2, ClusterRequestExecutor)> _requestExecutorsCache = new ConcurrentDictionary<string, (X509Certificate2, ClusterRequestExecutor)>();
```
Comment (ayende) [inline src/Raven.Server/Utils/ReplicationUtils.cs]:
```suggestion
        private static readonly ConcurrentDictionary<(string Url, X509Certificate2 Cert),  Lazy<ClusterRequestExecutor>> _requestExecutorsCache = new ConcurrentDictionary<(string Url, X509Certificate2 Cert),  Lazy<ClusterRequestExecutor>>();
```

File: src/Raven.Server/Utils/ReplicationUtils.cs [hunk 3]
```diff
@@ -13,6 +15,35 @@ namespace Raven.Server.Utils
...
+                _requestExecutorsCache.Remove(url, out _);
+                reqExecutor.Dispose();
+            }
+
+            if (_requestExecutorsCache.Count > 128) // clear from time to time
+            {
+                while (_requestExecutorsCache.Count > 0)
+                {
+                    _requestExecutorsCache.Remove(url, out var certAndRequestor);                    
+                    certAndRequestor.Item2.Dispose();
```
Comment (ayende) [inline src/Raven.Server/Utils/ReplicationUtils.cs]: This is probably not safe, what is we are currently using this?

File: src/Raven.Server/Utils/ReplicationUtils.cs [hunk 4]
```diff
@@ -13,6 +15,35 @@ namespace Raven.Server.Utils
...
+                }
+            }
+
+            reqExecutor = ClusterRequestExecutor.CreateForSingleNode(url, certificate);
+            _requestExecutorsCache[url] = (certificate, reqExecutor);
```
Comment (ayende) [inline src/Raven.Server/Utils/ReplicationUtils.cs]: Not concurrently safe

File: src/Sparrow/Json/JsonContextPoolBase.cs [hunk 5]
```diff
@@ -1,138 +1,69 @@
...
-            return ThreadId.ToString();
-        }
-    }
-
-    public abstract class JsonContextPoolBase<T> : ILowMemoryHandler, IDisposable 
+{    
+    public abstract class JsonContextPoolBase<T> : ILowMemoryHandler, IDisposable
         where T : JsonOperationContext
     {
+        public object ExternalState { get; set; }
```
Comment (ayende) [inline src/Sparrow/Json/JsonContextPoolBase.cs:15]: We don't need this

File: src/Raven.Server/Utils/ReplicationUtils.cs (hunk 4 above)
Comment (aviviadi) [reply src/Raven.Server/Utils/ReplicationUtils.cs]: Thread safety : https://github.com/ravendb/ravendb/pull/9871/commits/4c6e96b6306919736303fa7e2c9c0b2ef0438c07

File: src/Raven.Server/Utils/ReplicationUtils.cs (hunk 2 above)
Comment (aviviadi) [reply src/Raven.Server/Utils/ReplicationUtils.cs]: This needs calculating hash of string and object. Why not use 'string url' so we use string comperator and we have single certificate per url any how, so this way it is more efficient lookup. No?

**Comment by aviviadi** [comment]
oookey.. I understand from @karmeli87 there's might be use of url and few certificates. 
I will rebase and remove the cache commits, and will create general thread safe Executor cache.

**Comment by arekpalinski** [comment]
We're getting a lot of test failures on this PR. Sample builds:

https://jenkins.hibernatingrhinos.com/job/PR_Tests_v4.2/2370/
https://jenkins.hibernatingrhinos.com/job/PR_Tests_v4.2/2374/
https://jenkins.hibernatingrhinos.com/job/PR_Tests_v4.2/2373/

Timeout:
https://jenkins.hibernatingrhinos.com/job/PR_Tests_v4.2/2375/console

**Comment by aviviadi** [comment]
Closed and pull requested again at : https://github.com/ravendb/ravendb/pull/9882

**Pull request closed without merging**